# Solution (1)

## (i) Determine all the eigenvalues of A

In [ ]:
import numpy as np

A = np.array([
    [6, 1, 1, 1],
    [1, 5, 1, 1],
    [1, 1, 4, 1],
    [1, 1, 1, 6]
], dtype=np.float64)

eigenvalues, eigenvectors = np.linalg.eig(A)

eigenvalues # list of eigenvalues of A

array([8.40267883, 5.        , 4.31603086, 3.28129031])

## (ii) Determine Y (t) corresponding to the given initial value problem.

In [ ]:
D = np.diag(eigenvalues)
D # diagonal matrix similar to A

array([[8.40267883, 0.        , 0.        , 0.        ],
       [0.        , 5.        , 0.        , 0.        ],
       [0.        , 0.        , 4.31603086, 0.        ],
       [0.        , 0.        , 0.        , 3.28129031]])

In [ ]:
from sympy import Matrix, symbols

t = symbols('t')

D = Matrix(D.tolist())

Dt = D * t

expDt = Dt.exp()

expDt # e^tD

Matrix([
[1.0*exp(8.40267882952118*t),              0,                           0,                           0],
[                          0, 1.0*exp(5.0*t),                           0,                           0],
[                          0,              0, 1.0*exp(4.31603086087072*t),                           0],
[                          0,              0,                           0, 1.0*exp(3.28129030960808*t)]])

In [ ]:
P = eigenvectors
P_inv = np.linalg.inv(P)

P = Matrix(P.tolist())
P_inv = Matrix(P_inv.tolist())

y_0 = Matrix([[1], [1], [1], [1]])

y = P * expDt * P.inv() * y_0 # Here we used the result that y = P @ e^tD @ P^-1 @ y0

y # solution to the given IVP

Matrix([
[-0.0381495200542136*exp(3.28129030960808*t) - 0.0983550967727958*exp(4.31603086087072*t) + 1.33226762955019e-15*exp(5.0*t) + 1.13650461682701*exp(8.40267882952118*t)],
[ -0.091230090087172*exp(3.28129030960808*t) + 0.212864815427578*exp(4.31603086087072*t) - 8.87468518373638e-31*exp(5.0*t) + 0.878365274659594*exp(8.40267882952118*t)],
[ 0.233097079996577*exp(3.28129030960808*t) + 0.0511172289866793*exp(4.31603086087072*t) - 2.95822839457879e-31*exp(5.0*t) + 0.715785691016744*exp(8.40267882952118*t)],
[-0.0381495200542137*exp(3.28129030960808*t) - 0.0983550967727962*exp(4.31603086087072*t) - 1.27675647831893e-15*exp(5.0*t) + 1.13650461682701*exp(8.40267882952118*t)]])

## (iii) Whether the answer Y (t) is asymptotically stable? Justify your answer.

### Answer - y(t) is asymptotically unstable and the given system is an unstable system

For the given system, the real part of all eigenvalues of A is greater 1 which implies all components of y(t) tend to infinity as t tends to infinity. Therefore, y(t) is unstable asymptotically and the given system in an unstable system.

# Solution (2)

## (i) Use Gaussian Elimination method to solve this problem.

In [73]:
A = np.array([
    [6, 1, 1, 1],
    [1, 5, 1, 1],
    [1, 1, 4, 1],
    [1, 1, 1, 6]
], dtype=np.float64)

b = np.array([4, 3, 2, 1], dtype=np.float64)

x = np.linalg.solve(A, b)

x

array([ 0.55630252,  0.44537815,  0.2605042 , -0.04369748])

### Answer is x = [ 0.55630252,  0.44537815,  0.2605042 , -0.04369748]

## (ii) Jacobi iteration

In [19]:
L = np.array([
    [0, 0, 0, 0],
    [1, 0, 0, 0],
    [1, 1, 0, 0],
    [1, 1, 1, 0]
], dtype=np.float64)

D = np.array([
    [6, 0, 0, 0],
    [0, 5, 0, 0],
    [0, 0, 4, 0],
    [0, 0, 0, 6]
], dtype=np.float64)

U = np.array([
    [0, 1, 1, 1],
    [0, 0, 1, 1],
    [0, 0, 0, 1],
    [0, 0, 0, 0]
], dtype=np.float64)

In [20]:
D_inv = np.linalg.inv(D)
T = -D_inv @ (L + U)
T

array([[ 0.        , -0.16666667, -0.16666667, -0.16666667],
       [-0.2       ,  0.        , -0.2       , -0.2       ],
       [-0.25      , -0.25      ,  0.        , -0.25      ],
       [-0.16666667, -0.16666667, -0.16666667,  0.        ]])

In [24]:
b = np.array([4, 3, 2, 1], dtype=np.float64)
c = D_inv @ b
u_prev = np.array([0, 0, 0, 0], dtype=np.float64)

for k in range(0, 20):
    u_next = T @ u_prev + c
    u_prev = u_next

u_next

array([ 0.5562969 ,  0.44537169,  0.26049661, -0.0437031 ])

### Answer is u_20 = [ 0.5562969 ,  0.44537169,  0.26049661, -0.0437031 ]

# Solution (3)

Mistake - Work with B.T

### We use Perron-Frobenius Theorem

In [25]:
import numpy as np

B = np.array([
    [0, 0.5, 0.3, 0.6],
    [0.3, 0, 0.4, 0.1],
    [0.3, 0.4, 0, 0.3],
    [0.4, 0.1, 0.3, 0]
], dtype=np.float64)

In [26]:
B @ B

array([[0.48, 0.18, 0.38, 0.14],
       [0.16, 0.32, 0.12, 0.3 ],
       [0.24, 0.18, 0.34, 0.22],
       [0.12, 0.32, 0.16, 0.34]])

### B^2 has all entries non-zero, therefore we can apply Perron-Frobenius theorem to get the steady state probability vector which is the eigenvector corresponding to eigenvalue 1.

In [27]:
eigenvalues, eigenvectors = np.linalg.eig(B)

print(f'eigenvalues:\n{eigenvalues}\n')
print(f'eigenvectors:\n{eigenvectors}\n')

eigenvalues:
[ 1.         -0.05857864 -0.6        -0.34142136]

eigenvectors:
[[ 0.62258139  0.27059805 -0.76200076 -0.65328148]
 [ 0.42606827 -0.65328148  0.38100038 -0.27059805]
 [ 0.48867422 -0.27059805 -0.12700013  0.65328148]
 [ 0.43824165  0.65328148  0.50800051  0.27059805]]



### 1st column of eigenvectors is the steady state probability vector - we can normalize it to obtain a probability vector

In [28]:
u_star = eigenvectors[:, 0]
u_star

array([0.62258139, 0.42606827, 0.48867422, 0.43824165])

In [29]:
s = np.sum(u_star)
u_star /= s
u_star

array([0.31514085, 0.21566901, 0.24735915, 0.22183099])

### Final answer - a1 is most visited and a2 is least visited

probability vector u_star is [0.31514085, 0.21566901, 0.24735915, 0.22183099]

# Solution (4)

Wrong

Can use formula for - Matrix Functions Via Jordan Canonical Form

or use

```py
from scipy.linalg import sinm, cosm
```

In [30]:
import numpy as np

M = np.array([
    [4, 1],
    [-1, 6]
], dtype=np.float64)

eigenvals, eigenvects = np.linalg.eig(M)

print(f'eigenvals:\n{eigenvals}\n')
print(f'eigenvects:\n{eigenvects}\n')

eigenvals:
[5. 5.]

eigenvects:
[[ 0.70710678 -0.70710678]
 [ 0.70710678 -0.70710678]]



### So, A is not diagonalizable. We go for JCF for which we'll use sympy

In [ ]:
from sympy import Matrix

M = Matrix([
    [4, 1],
    [-1, 6]
])

P, J = M.jordan_form()

Matrix([
[5, 1],
[0, 5]])

In [41]:
P

Matrix([
[-1, 1],
[-1, 0]])

In [42]:
J

Matrix([
[5, 1],
[0, 5]])

### There's no function in sympy to compute sin and cos of the 2x2 Jordan block directly. Instead, we use taylor series expansion of sin and cos.

Write J as lamda*I + N, N is nilpotent with m = 2 (N^m = O)

In [ ]:
P = np.array([
    [-1, 1],
    [-1, 0]
], dtype=np.float64)

P_inv = np.linalg.inv(P)

# for sin
# A1 is the nilpotent matrix part in the 2x2 Jordan block
A1 = np.array([
    [0, 1],
    [0, 0]
], dtype=np.float64)

sin_D = np.diag(np.sin(eigenvals)) + A1     # by expanding the taylor series of sin

# for cos
cos_D = np.diag(np.cos(eigenvals)) + np.eye(2)      # by expanding the taylor series of cos

In [47]:
sin_M = P @ sin_D @ P_inv

sin_M

array([[-1.95892427,  1.        ],
       [-1.        ,  0.04107573]])

In [48]:
cos_M = P @ cos_D @ P_inv

cos_M

array([[1.28366219, 0.        ],
       [0.        , 1.28366219]])

In [49]:
sin_M @ sin_M + cos_M @ cos_M

array([[ 4.48517292, -1.91784855],
       [ 1.91784855,  0.64947582]])

# Solution (5)

Missed

```py
sorted_idx = np.argsort(eigenvalues)[::-1]
eigenvalues = eigenvalues[sorted_idx]
eigenvectors = eigenvectors[:, sorted_idx]
```

In [117]:
X = np.loadtxt('Question5.csv', delimiter=',')
X       # given data matrix

array([[7.5, 7.5, 7. , 5. , 8. ],
       [7.5, 6. , 5.5, 6. , 7.5],
       [7. , 7.5, 7. , 5. , 7. ],
       [6.5, 5. , 6. , 5. , 7. ],
       [9. , 8. , 7.5, 6. , 8. ],
       [6.5, 5. , 6. , 3.5, 7. ],
       [7.5, 7.5, 5.5, 6. , 7. ],
       [7. , 7. , 6. , 5. , 7.5],
       [7.5, 6. , 6.5, 6. , 8. ],
       [8.5, 8. , 7. , 6. , 8. ]])

In [118]:
X_c = X - np.mean(X, axis=0)

K = (X_c.T @ X_c) / 9

eigenvalues, eigenvectors = np.linalg.eigh(K)

W = eigenvectors[:, :4]
W       # all 5 principal components

array([[-4.99691872e-01, -6.03377620e-01,  3.61195782e-01,
        -1.69706480e-01],
       [ 1.71276697e-01, -3.72104519e-02, -6.13034085e-01,
         3.06898100e-01],
       [-1.83781750e-01,  5.55439645e-01,  5.29380480e-01,
         5.36005136e-01],
       [-6.51638316e-04,  5.21532390e-01,  1.61673138e-02,
        -7.66457905e-01],
       [ 8.28972961e-01, -2.32468485e-01,  4.61759399e-01,
        -4.74767386e-02]])

In [119]:
mu = np.mean(X, axis=0)
mu      # mean vector

array([7.45, 6.75, 6.4 , 5.35, 7.5 ])

In [120]:
# x is the given rating with 3rd component missing
# first center x using the mean vector mu
x = np.array([7 - 7.45, 6 - 6.75, 6 - 5.35, 7.5 - 7.5 ])
x

array([-0.45, -0.75,  0.65,  0.  ])

In [ ]:
# x is in the best 4D subspace
# So, we can write x as x = Wz for some z
# We'll find z

# Compute W without 3rd row
# as 3rd row of W is used to get the 3rd component of x (which is to be calculated)
A = []
for i in range(5):
    if i == 2:
        continue
    a = W[i]
    A.append(list(a))

A = np.array(A)
A

array([[-4.99691872e-01, -6.03377620e-01,  3.61195782e-01,
        -1.69706480e-01],
       [ 1.71276697e-01, -3.72104519e-02, -6.13034085e-01,
         3.06898100e-01],
       [-6.51638316e-04,  5.21532390e-01,  1.61673138e-02,
        -7.66457905e-01],
       [ 8.28972961e-01, -2.32468485e-01,  4.61759399e-01,
        -4.74767386e-02]])

In [ ]:
# solve Wz = x for z
z = np.linalg.solve(A, x)
z

array([-0.21058306,  1.56494356,  1.19079709,  0.24209806])

In [ ]:
W @ z + mu  # x = Wz + mu   adding back the mean vector

array([7.        , 6.        , 8.06808356, 6.        , 7.5       ])

## Answer is - value of * is 8.0680835